### PR TITLE
R ovh domains zone

### PR DESCRIPTION
**Pull request by remijouannet**
Hello @yanndegat 

As you ask in the previous pr https://github.com/hashicorp/terraform/pull/14775
i moved my pr in this project,

here is my little readme again so we can continue to review my pr here 

howto

if you want to use the OVH API you have to generate:

An Application Key
An Application Secret Key
A Consumer Key

you can generate all three on this page
https://eu.api.ovh.com/createToken/

the following rights are needed for this plugin

GET : /domain/zone/*
PUT : /domain/zone/*
POST : /domain/zone/*
DELETE : /domain/zone/*
GET : /me

ovh.tf example

provider "ovh" {
application_key = "azrzqrgqvvdsfgsfffgc"
application_secret = "aztfqsqfsdcsdqezrfdvcx"
consumer_key = "aergfvdsrgtfbvcretfgd"
}

resource "ovh_domain_zone_record" "test" {
zone = "testdemo.ovh"
subDomain = "test"
fieldType = "A"
ttl = "3600"
target = "0.0.0.0"
}

testacc

export OVH_APPLICATION_KEY="arqzergfazfeef"
export OVH_APPLICATION_SECRET="aertedytfghftyhytghbgfv"
export OVH_CONSUMER_KEY="tyfghjhuyghvbjhuytghbjygh"
export OVH_ZONE="testdemo.ovh"

then use the makefile : make testacc TEST=./builtin/providers/ovh

### CODE REVIEW

**Comment by yanndegat** [comment]
hi @remijouannet 

FYI, i'm currently working with ovh to get a free ovh zone so that we can merge this PR and run the tests on HC infrastructure.

**Comment by remijouannet** [comment]
Hi @yanndegat , 
no problem i'm not in the hurry, 
you also can buy really cheap domain with the ext .ovh

**Comment by yanndegat** [comment]
yes but i'm not able to ask hashicorp to do so

-yann

File: ovh/resource_ovh_dns_record.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+			},
+			"zone": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"target": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"ttl": {
```
Comment (grubernaut) [inline ovh/resource_ovh_dns_record.go]: Is there a reason `ttl` needs to be a `TypeString` and not a `TypeInt`? Would save you from having to call `strconv` during Create.

File: ovh/resource_ovh_dns_record.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+	provider := meta.(*Config)
+
+	// Create the new record
+	newRecord := &NewRecord{
+		FieldType: d.Get("fieldType").(string),
+		SubDomain: d.Get("subDomain").(string),
+		Target:    d.Get("target").(string),
+	}
+
+	newRecord.Ttl, _ = strconv.Atoi(d.Get("ttl").(string))
```
Comment (grubernaut) [inline ovh/resource_ovh_dns_record.go]: If `ttl` has to be a string, dropping the error here.

File: ovh/resource_ovh_dns_record.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+		newRecord,
+		&resultRecord,
+	)
+
+	if err != nil {
+		return fmt.Errorf("Failed to create OVH Record: %s", err)
+	}
+
+	d.SetId(strconv.Itoa(resultRecord.Id))
+	d.Set("id", strconv.Itoa(resultRecord.Id))
```
Comment (grubernaut) [inline ovh/resource_ovh_dns_record.go]: duplicate set to `id`

File: ovh/resource_ovh_dns_record.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+		fmt.Sprintf("/domain/zone/%s/record", d.Get("zone").(string)),
+		newRecord,
+		&resultRecord,
+	)
+
+	if err != nil {
+		return fmt.Errorf("Failed to create OVH Record: %s", err)
+	}
+
+	d.SetId(strconv.Itoa(resultRecord.Id))
```
Comment (grubernaut) [inline ovh/resource_ovh_dns_record.go]: Can use `fmt.Sprintf("%d", resultRecord.Id))` here

File: ovh/resource_ovh_dns_record.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+	d.Set("id", strconv.Itoa(resultRecord.Id))
+	log.Printf("[INFO] OVH Record ID: %s", d.Id())
+
+	return resourceOVHRecordRead(d, meta)
+}
+
+func resourceOVHRecordRead(d *schema.ResourceData, meta interface{}) error {
+	provider := meta.(*Config)
+
+	recordID, err := strconv.Atoi(d.Id())
```
Comment (grubernaut) [inline ovh/resource_ovh_dns_record.go]: Don't believe you actually need this conversion back to an `int` here.

File: ovh/resource_ovh_dns_record.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+	if err != nil {
+		return fmt.Errorf("Error converting Record ID: %s", err)
+	}
+
+	record := Record{}
+	err = provider.OVHClient.Get(
+		fmt.Sprintf("/domain/zone/%s/record/%d", d.Get("zone").(string), recordID),
```
Comment (grubernaut) [inline ovh/resource_ovh_dns_record.go]: As you can just use `%s` for the `id` here as you're building a string.

File: ovh/resource_ovh_dns_record.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+	if err != nil {
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("id", record.Id)
+	d.Set("zone", record.Zone)
+	d.Set("fieldType", record.FieldType)
+	d.Set("subDomain", record.SubDomain)
+	d.Set("ttl", strconv.Itoa(record.Ttl))
```
Comment (grubernaut) [inline ovh/resource_ovh_dns_record.go]: Same with `ttl` here

File: ovh/resource_ovh_dns_record.go [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+	d.Set("target", record.Target)
+
+	return nil
+}
+
+func resourceOVHRecordUpdate(d *schema.ResourceData, meta interface{}) error {
+	provider := meta.(*Config)
+
+	recordID, err := strconv.Atoi(d.Id())
```
Comment (grubernaut) [inline ovh/resource_ovh_dns_record.go]: Can leave this as a string

File: ovh/resource_ovh_dns_record.go [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+		fmt.Sprintf("/domain/zone/%s/record/%d", d.Get("zone").(string), recordID),
+		&record,
+	)
+
+	if err != nil {
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("id", record.Id)
```
Comment (grubernaut) [inline ovh/resource_ovh_dns_record.go]: `record.Id` is an `int`, can't use it for `d.Set()`. Need to use `fmt.Sprintf("%d", record.Id)` or `strconv`

File: ovh/resource_ovh_dns_record.go [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+	return resourceOVHRecordRead(d, meta)
+}
+
+func resourceOVHRecordDelete(d *schema.ResourceData, meta interface{}) error {
+	provider := meta.(*Config)
+
+	log.Printf("[INFO] Deleting OVH Record: %s.%s, %s", d.Get("zone").(string), d.Get("subDomain").(string), d.Id())
+
+	recordID, err := strconv.Atoi(d.Id())
```
Comment (grubernaut) [inline ovh/resource_ovh_dns_record.go]: Can skip type conversion here, and continue using as a string on line 174

**Comment by remijouannet** [comment]
thx for the review, will look at it as soon as I have free time for it

File: ovh/resource_ovh_dns_record.go (hunk 4 above)
Comment (remijouannet) [inline ovh/resource_ovh_dns_record.go]: is there really a difference between using Sprintf and strconv ?

File: ovh/resource_ovh_dns_record.go (hunk 3 above)
Comment (remijouannet) [inline ovh/resource_ovh_dns_record.go]: @grubernaut as i understand terraform generate internally an ID for every ressource, 
the OVH api return a ID for each DNS entry so I force the internal terraform ID to be the ID return by the API
as this id is also part of the resource i'm setting it in the resource data schema

I don't know if i'm clear here and if this right way to do it

**Comment by remijouannet** [comment]
@yanndegat @grubernaut i made a few change following your comment, ttl and id type mostly

File: ovh/resource_ovh_dns_record.go (hunk 9 above)
Comment (remijouannet) [inline ovh/resource_ovh_dns_record.go]: i've changed the type of id in the data schema

**Comment by remijouannet** [comment]
@yanndegat @grubernaut Hello, any news ?

**Comment by grubernaut** [comment]
Hey @remijouannet, sorry for the delay. Looking now.

File: ovh/resource_ovh_dns_record.go (hunk 1 above)
Comment (grubernaut) [inline ovh/resource_ovh_dns_record.go]: @remijouannet ping ^

Comment (grubernaut) [inline ovh/resource_ovh_dns_record.go]: Nevermind, saw you changed it to `TypeInt` :smile:

File: website/docs/r/ovh_domain_zone_record.html.markdown [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+``hcl
+# Add a record to a sub-domain
+resource "ovh_domain_zone_record" "test" {
+    zone = "testdemo.ovh"
+    subDomain = "test"
+    fieldType = "A"
+    ttl = "3600"
+    target = "0.0.0.0"
+}
+``
```
Comment (grubernaut) [inline website/docs/r/ovh_domain_zone_record.html.markdown]: Missing '`' here

**Comment by grubernaut** [comment]
Hey @remijouannet, looks good, thanks for the contribution! 
Just need a `zone` to add for our acceptance tests, and this can be merged.

**Comment by remijouannet** [comment]
@grubernaut still in discussion with OVH for a free domain for your acceptance tests ?

**Comment by grubernaut** [comment]
I believe @yanndegat was in conversations with them, but if they are unable to provide one to us for free for acceptance testing, or require our input to do so, we can help out there. If they're fully unable to provide one to us for free testing, @randomcamel should be able to get us a paid account for it.

**Comment by remijouannet** [comment]
@grubernaut @randomcamel  do you already have an account for the openstack part ? doing all the test with the same account should be more convenient, an .ovh domain is very cheap like 3e/year, it should be enought for the acceptance tests

**Comment by grubernaut** [comment]
@remijouannet Aye, we currently have an OVH account that we test with, just need the person with the credit card (**cough** **cough** @randomcamel) to add a domain :smile:

**Comment by yanndegat** [comment]
hi currently on holidays
news from ovh is that they can provide something for this but i'll have to go back through another round of emails starting september. If you guys from hashicorp can purchase an ovh domain for our tests purpose it would be great. if thinks ovh domain is 1$ a year.

**Comment by grubernaut** [comment]
Hey @yanndegat, 

Thanks for the patience in getting this in! We're currently working alongside OVH to get a domain for testing purposes, but are running into some issues there. Stay tuned for more!

**Comment by remijouannet** [comment]
hey @yanndegat @grubernaut just want to know the status of the discussion with OVH

**Comment by remijouannet** [comment]
heyhey @yanndegat @grubernaut just passing by to remind you the existence of this PR

**Comment by yanndegat** [comment]
and from my side, i work for ovh now, and i should be able to help for this
setup, and get a voucher for this purpose.


-yann

**Comment by tounefr** [comment]
This feature would be very usefull.
However, i've noticed the domain needs to be refreshed after each update:
https://api.ovh.com/console/#/domain/zone/%7BzoneName%7D/record#POST
> Create a new DNS record (Don't forget to refresh the zone)

Refresh request:
https://api.ovh.com/console/#/domain/zone/%7BzoneName%7D/refresh#POST

https://github.com/remijouannet/terraform-provider-ovh/pull/1

**Comment by remijouannet** [comment]
hello @tounefr fix in this pull request as well

**Comment by remijouannet** [comment]
Hello @catsby @randomcamel, 
it's been a 5 month since this PR is open, 
the code has been reviewed, fix and validate, 
can you please get a OVH domain for the acceptance tests (@yanndegat can help us for that)
so this PR can be merge

**Comment by yanndegat** [comment]
hi,

@catsby @randomcamel

i can get you a voucher for this

if you give me the ovh account that is registered for hashicorp i can even
make all the setup for you

regards

-yann

**Comment by smola** [comment]
I'm using this branch after rebase with latest master and it's working perfectly so far.

**Comment by remijouannet** [comment]
Hi @smola, can you be more specific ?

**Comment by tombuildsstuff** [comment]
I've been working with @yanndegat to allow the CI system to be able to run these tests - this has now been completed and the tests can run (and pass!):

screen shot 2017-12-05 at 12 16 35

**Comment by remijouannet** [comment]
@tombuildsstuff @yanndegat  awesome thx for your work

**Comment by yanndegat** [comment]
thanks for your time and your patience!

and thank you once again @tombuildsstuff

-yann

**Pull request merged**